stm32 LPTIM clock with prescaler to adjust the TICKS_PER_SEC

FRASTM opened the pull request:
When the LPTIM is enabled for the lowPower (PM) purpose, the SYS_CLOCK_TICKS_PER_SEC is set to a lower value (than default 10000)

- 4000 if LPTIM is clocked by LSI (32000Hz)
- 4096 if LPTIM is clocked by LSE (32768Hz)

However when the lptim has a <prescaler> that divides the clock source , this SYS_CLOCK_TICKS_PER_SEC is even too high, resulting in numerous system wakeup during sleep mode (no more sleeping indeed)
Consequently the SYS_CLOCK_TICKS_PER_SEC is set to a value depending 
 on the clock input LSE/LSI and the clock prescaler property set in the DTS
- 1000 if LPTIM is clocked by LSI (32000Hz)  divided by 32 
- 1024 if LPTIM is clocked by LSE (32768Hz)  divided by 32 
- 2000 if LPTIM is clocked by LSI (32000Hz)  divided by 16 
- 2048 if LPTIM is clocked by LSE (32768Hz)  divided by 16

Removing SYS_CLOCK_TICKS_PER_SEC in the board config and adding
Warning build message to check the coherency between
- clock source definition (LSI/LSE) and the board DTS (lptim node <clocks> property)
- SYS_CLOCK_TICKS_PER_SEC and the LPTIM clock ratio (lptim node <st,prescaler> property) 

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/61099

FRASTM committed:
samples: boards: stm32 low power blinky for stm32 boards

Set the prescaler for the stm32_lp_tick_source lptim node.
When the LPTIM is clocked by the LSE with a prescaler
of 16 (lptim freq at 2048Hz) expecting 2048 for the TICKS_PER_SEC.
When the LPTIM is clocked by the LSE with a prescaler
of 32 (lptim freq at 1024Hz) expecting 1024 for the TICKS_PER_SEC.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -46,6 +46,10 @@ static const struct stm32_pclken lptim_clk[] = STM32_DT_INST_CLOCKS(0);
 
 static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
 
+/* For tick accuracy, a specific tick to freq ratio is expected */
+/* This check assumes LSI@32KHz or LSE@32768Hz */
```
Comment applies to a check while the next line is a definition. This is confusing

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -46,6 +46,10 @@ static const struct stm32_pclken lptim_clk[] = STM32_DT_INST_CLOCKS(0);
 
+/* For tick accuracy, a specific tick to freq ratio is expected */
+/* This check assumes LSI@32KHz or LSE@32768Hz */
+#define LPTIM_CLOCK_RATIO DT_PROP(DT_DRV_INST(0), st_prescaler)
```
Since you're mentioning `freq ratio` just before, naming the C value equivalent of `st-prescaler` as `LPTIM_CLOCK_RATIO` is a bit disturbing.
What about `LPTIM_PRESCALER`? Now that there is actually an important ratio notion in the driver, it may help the reader.

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -61,11 +65,41 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
-static uint32_t lptim_clock_freq = KHZ(32);
-static int32_t lptim_time_base;
+#if (CONFIG_STM32_LPTIM_CLOCK_LSI)
+static uint32_t lptim_clock_freq = KHZ(32); /* Force STM32_SRC_LSI */
```
I don't know how we were handling this previously but on some series, LSI could be different from 32K
L0, L1, F0, F1, F3
Not all of them use LPTIM driver (Only L0 today), but a comment on this case would be welcome.

erwango replied:
L0 uses LPTIM

FRASTM replied:
> We could probably use `STM32_LSI_FREQ` defined in stm32_clock_control.h. That would take care of cases where LSI is not 32 kHz.

in case  the LSI is 37kHz like stm32l0x, the lptim_clock_freq  is set to 32kHz so this cannot be STM32_LSI_FREQ

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -61,11 +65,41 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
-#define LPTIM_CLOCK_RATIO DT_PROP(DT_DRV_INST(0), st_prescaler)
+/* Kconfig defines the clock source as LSI : check coherency with DTS */
+#if (DT_INST_CLOCKS_CELL_BY_IDX(0, 1, bus) != STM32_SRC_LSI)
+#warning CONFIG_STM32_LPTIM_CLOCK_LSI requires STM32_SRC_LSI (DTS)
```
```suggestion
#warning CONFIG_STM32_LPTIM_CLOCK_LSI requires STM32_SRC_LSI defined as LPTIM domain clock
```

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -61,11 +65,41 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
+
+/* Kconfig defines the clock source as LSE : check coherency with DTS */
+#if (DT_INST_CLOCKS_CELL_BY_IDX(0, 1, bus) != STM32_SRC_LSE)
+#warning CONFIG_STM32_LPTIM_CLOCK_LSE requires STM32_SRC_LSE (DTS)
```
```suggestion
#warning CONFIG_STM32_LPTIM_CLOCK_LSE requires STM32_SRC_LSE defined as LPTIM domain clock
```

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -61,11 +65,41 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
+#endif /* STM32_SRC_LSI */
+
+/* Kconfig should set the adapted CONFIG_SYS_CLOCK_TICKS_PER_SEC */
+#if !defined(CONFIG_STM32_LPTIM_TICK_FREQ_RATIO_OVERRIDE)
```
Is there a real interest in `CONFIG_STM32_LPTIM_TICK_FREQ_RATIO_OVERRIDE`, now that we understand all constraints related to LPTIM configuration? 
```
	  For LPTIM configuration, a specific tick freq is advised
	  depending on LPTIM input clock:
	  - LSI(32KHz): 4000 ticks/sec
	  - LSE(32678): 4096 ticks/sec
	  To prevent misconfigurations, a dedicated check is implemented
	  in the driver.
	  This options allows to override this check
```
If we need to keep it (which seems to be the case for ZTEST), we might want to update the definition

gautierg-st replied:
I'm also wondering what is the benefit of all these checks since a default SYS_CLOCK_TICKS_PER_SEC (that verifies the conditions) is already provided.

FRASTM replied:
Even if the SYS_CLOCK_TICKS_PER_SEC is given by the soc/arm/st_stm32/common/Kconfig.defconfig.series
an application can still set it in its prj.conf

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -61,11 +65,41 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
+/* Kconfig should set the adapted CONFIG_SYS_CLOCK_TICKS_PER_SEC */
+#if !defined(CONFIG_STM32_LPTIM_TICK_FREQ_RATIO_OVERRIDE)
+#if ((CONFIG_SYS_CLOCK_TICKS_PER_SEC > 32000/LPTIM_CLOCK_RATIO) || \
+	(CONFIG_SYS_CLOCK_TICKS_PER_SEC > 4000))
```
I'm not sure how both conditions interact here:
if LPTIM_CLOCK_RATIO = 2, CONFIG_SYS_CLOCK_TICKS_PER_SEC = 8000 is valid according to the first check, and inline with the comment, but `CONFIG_SYS_CLOCK_TICKS_PER_SEC > 4000` would generate the warning

FRASTM replied:
once LPTIM is enabled  CONFIG_SYS_CLOCK_TICKS_PER_SEC  is expected to be  equal to 4000 or lower if prescaler is >= 16

erwango replied:
Yes, but we're looking for a rule working in all cases. This is not the case of the condition you wrote.

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -61,11 +65,41 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
+#if !defined(CONFIG_STM32_LPTIM_TICK_FREQ_RATIO_OVERRIDE)
+#if ((CONFIG_SYS_CLOCK_TICKS_PER_SEC > 32000/LPTIM_CLOCK_RATIO) || \
+	(CONFIG_SYS_CLOCK_TICKS_PER_SEC > 4000))
+#warning It is recommended to set the tick freq to 32000 / prescaler
```
Warning is generated on a "lower than" condition, so warning message should be coherent with that.
For instance:
```
It is recommended to set the tick freq value lower than 32000 / prescaler
```


But in the end, I think it would be easier to understand if we have a warning about "tick per clock cycle":
(if possible, express the check in tick per clock cycle:)
```
#if (CONFIG_SYS_CLOCK_TICKS_PER_SEC * LPTIM_CLOCK_RATIO / CONFIG_STM32_LPTIM_CLOCK_LSE > 1)
```
Then write the warning using the same notion
```
#warning Configuration requests more than 1 tick per clock cycle. This is not valid.
```

FRASTM replied:
Another possibility is to complete the soc/arm/st_stm32/common/Kconfig.defconfig.series with 
```
if STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSE
config SYS_CLOCK_TICKS_PER_SEC
	default 4096 if "$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),clock-prescaler)" < 16
	default 2048 if "$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),clock-prescaler)" = 16
	default 1024 if "$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),clock-prescaler)" = 32
```

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -61,11 +65,41 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
+#endif /* STM32_SRC_LSE */
+
+/* Kconfig should set the adapted CONFIG_SYS_CLOCK_TICKS_PER_SEC */
+#if !defined(CONFIG_STM32_LPTIM_TICK_FREQ_RATIO_OVERRIDE)
+#if ((CONFIG_SYS_CLOCK_TICKS_PER_SEC > 32768/LPTIM_CLOCK_RATIO) || \
+	(CONFIG_SYS_CLOCK_TICKS_PER_SEC > 4096))
+#warning It is recommended to set the tick freq to 32768 / prescaler
+#endif
+#endif /* !CONFIG_STM32_LPTIM_TICK_FREQ_RATIO_OVERRIDE */
```
Same comment than with LSI.
If possible a common condition could be used (using an intermediate #define)

erwango commented on samples/boards/stm32/power_mgmt/blinky/boards/nucleo_wb55rg.conf:
```diff
@@ -0,0 +1,9 @@
+# with a LPTIM1 prescaler of <32>, set the CONFIG_SYS_CLOCK_TICKS_PER_SEC to 1024
+CONFIG_SYS_CLOCK_TICKS_PER_SEC=1024
+
+# also set the CONFIG_STM32_LPTIM_CLOCK_LSE
+# to match the <clocks> property of the lptim1 node in the DTS
+CONFIG_STM32_LPTIM_CLOCK_LSE=y
```
If LSE is configured in board.dts. `CONFIG_STM32_LPTIM_CLOCK_LSE=y` should also be part of default board configuration. 
Which is the case based on previous commit. This line could be cleaned up.

FRASTM replied:
see  https://github.com/zephyrproject-rtos/zephyr/pull/64067

erwango reviewed (changes requested):
I think this is a good occasion to remove the deprecation message on "#if DT_INST_NUM_CLOCKS(0) == 1
#warning Kconfig for LPTIM source clock (LSI/LSE) is deprecated, use device tree."

FRASTM committed:
soc: arm: stm32 SYS_CLOCK_TICKS_PER_SEC config with STM32_LPTIM_TIMER

Configure the SYS_CLOCK_TICKS_PER_SEC directly from the
DTS st-prescaler property of the lptim node
aka stm32_lp_tick_source

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32 boards with lptimer set node stm32_lp_tick_source

Change the name of the node for the lptim used as lowpower
tick source to stm32_lp_tick_source.
Once enabled, this node is known as stm32_lp_tick_source
That will avoid naming the node lptim1 or lptim2 or lptim, etc.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Rebase on 68365d5b67057c8e960a5168c2cf99f4b8c457c1
Update the SYS_CLOCK_TICKS_PER_SEC  in the soc/arm/st_stm32/common/Kconfig.defconfig.series
to retrieve the prescaler from the deviceTree. 
Requires to define the lptim prescaler property as <clock-prescaler>  (instead of "st,prescaler") 
because it is not accepted by the macro `$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),clock-prescaler)`
Add a  generic name `stm32_lp_tick_source`  to the lptim node (= the lptim enabled for low power tick counter)
  a board overlay is now defining the prescaler with :
```
&stm32_lp_tick_source {
	clock-prescaler = <32>;
};
```

erwango commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -24,9 +26,21 @@ config LOG_BACKEND_SWO_REF_FREQ_HZ
 config SYS_CLOCK_TICKS_PER_SEC
-	default 4096 if STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSE
-	default 4000 if STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSI
+	default 4096 if "$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),clock-prescaler)" < 16
```
Before changing the binding (which will have wide impacts) let's try to define a macro :
```
DT_ST_PRESCALER := st,prescaler
```
then use it as follows:
```
	default 4096 if "$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),$(DT_ST_PRESCALER))" < 16
```

FRASTM replied:
yes, exactly, now I have the macro ,-)

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -61,11 +64,63 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
+#if (CONFIG_STM32_LPTIM_CLOCK_LSI)
+/*
+ * Most of the STM32_SERIES have a LSI @32000Hz,
+ * this frequency might differ on some that do not yet use LPTIM
```
LPTIM is available on L0.
There is some dedicated code around it. See https://github.com/zephyrproject-rtos/zephyr/blame/main/drivers/timer/stm32_lptim_timer.c#L368

erwango replied:
I guess this comment is not relevant at this place anymore.

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -61,11 +64,63 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
+#if !defined(CONFIG_STM32_LPTIM_TICK_FREQ_RATIO_OVERRIDE)
+/* For prescaler = 1,2,4,8 CONFIG_SYS_CLOCK_TICKS_PER_SEC is <= 4000 */
+#if ((CONFIG_SYS_CLOCK_TICKS_PER_SEC > 32000/8) && \
+	(LPTIM_CLOCK_RATIO < 16))
```
Please change `LPTIM_CLOCK_RATIO` to `LPTIM_PRESCALER`.
`LPTIM_CLOCK_RATIO` is quite confusing, specially when used in this context.

FRASTM replied:
LPTIM_CLOCK_RATIO replaced by  LPTIM_PRESCALER  in a dedicated commit

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -61,11 +64,63 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
+
+/* Kconfig is setting CONFIG_SYS_CLOCK_TICKS_PER_SEC adapted to the clock ratio */
+#if !defined(CONFIG_STM32_LPTIM_TICK_FREQ_RATIO_OVERRIDE)
+/* For prescaler = 1,2,4,8 CONFIG_SYS_CLOCK_TICKS_PER_SEC is <= 4000 */
```
/* For prescaler = 1,2,4,8 CONFIG_SYS_CLOCK_TICKS_PER_SEC <= 4000 is advised */

FRASTM replied:
comment changed

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -61,11 +64,63 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
+#endif
+/*
+ * For prescaler = 16, set CONFIG_SYS_CLOCK_TICKS_PER_SEC to <= 2000
+ * For prescaler >= 32, set CONFIG_SYS_CLOCK_TICKS_PER_SEC to <= 1000
```
Is that possible to have prescaler > 32 ?

FRASTM replied:
yes, according to the dts/bindings/timer/st,stm32-lptim.yaml

FRASTM replied:
extending the value of CONFIG_SYS_CLOCK_TICKS_PER_SEC when prescaler is 64 or 128

erwango commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -24,9 +27,21 @@ config LOG_BACKEND_SWO_REF_FREQ_HZ
 endif # LOG_BACKEND_SWO
 
 # set the tick per sec as a divider of the LPTIM clock source
+if STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSE
```
What about using `depends on STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSE` instead of  `if` ?

FRASTM replied:
same result, changing to "depends on"

erwango commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -24,9 +27,21 @@ config LOG_BACKEND_SWO_REF_FREQ_HZ
+
+endif # STM32_LPTIM_TIMER
+
+if STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSI
```
What about using `depends on STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSI` ?

FRASTM replied:
same result, changing to "depends on"

FRASTM committed:
drivers: timer: stm32 lptimer clock prescaler property

Rename to LPTIM_PRESCALER, the <st,property> prescaler of
the stm32 LPTimer. This commit gives better readability than
LPTIM_CLOCK_RATIO.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

gautierg-st commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -25,8 +28,20 @@ endif # LOG_BACKEND_SWO
 config SYS_CLOCK_TICKS_PER_SEC
-	default 4096 if STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSE
-	default 4000 if STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSI
+	default 4096 if "$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),$(DT_ST_PRESCALER))" < 16
```
Maybe add a comment why we don't go higher than 4096 for all prescaler < 16.

gautierg-st commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -25,8 +28,20 @@ endif # LOG_BACKEND_SWO
+	default 1024 if "$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),$(DT_ST_PRESCALER))" = 32
+	default  512 if "$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),$(DT_ST_PRESCALER))" = 64
+	default  256 if "$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),$(DT_ST_PRESCALER))" = 128
+	depends on STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSE
```
If possible, `STM32_LPTIM_CLOCK_LSI` and `STM32_LPTIM_CLOCK_LSE` should be derived automatically from the kernel clock source of the LPTIM defined in the dts.
That way, we could really get rid of the old way of setting the source clock.
You'll probably need to set the value of `STM32_SRC_LSI` and `STM32_SRC_LSE` to be identical across all series (but they're just an arbitrary index, so that shouldn't be a problem).

FRASTM replied:
The Kconfig should get the STM32_LPTIM_CLOCK_LSE or STM32_LPTIM_CLOCK_LSI directly from the device Tree. However there is no $(dt_inst_clocks_cell_by_idx(0, 1, bus)  macro available to retrieve the <clocks> property from the deviceTree.

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -61,11 +61,60 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
-#define LPTIM_CLOCK_RATIO DT_PROP(DT_DRV_INST(0), st_prescaler)
+#define LPTIM_PRESCALER DT_PROP(DT_DRV_INST(0), st_prescaler)
+
+#if (CONFIG_STM32_LPTIM_CLOCK_LSI)
```
I understand your comment, but I propose to have another PR for this change :  get the LPTIM clock source index from clock source selected in the DTS of the stm32_lp_tick_source node.

FRASTM commented:
rebase on 15cbe0f890a26b9c4d98a528293f9cf32588c2b9
Update the migration guide to zephyr 3.6

erwango commented on samples/boards/stm32/power_mgmt/blinky/boards/nucleo_wb55rg.overlay:
```diff
@@ -1,9 +1,9 @@
-
-&lptim1 {
+&stm32_lp_tick_source {
 	st,prescaler = <32>;
```
I thought that st,prescaler = <32>; is too much.

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -61,11 +61,60 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
+#endif /* !CONFIG_STM32_LPTIM_TICK_FREQ_RATIO_OVERRIDE */
+
+#elif (CONFIG_STM32_LPTIM_CLOCK_LSE)
+static uint32_t lptim_clock_freq = 32768; /* Force STM32_SRC_LSE */
```
Is that required ?

FRASTM replied:
` static uint32_t lptim_clock_freq = STM32_LPTIM_CLOCK`

erwango commented:
@FRASTM Do you think we can replace the whole prescaler/frequency Kconfig based check by a runtime version ?
I expect a runtime check to be much more simple. Drawback is that misconfiguration would be discovered later, but it doesn't prevent to provide a clear message to the user.

Also, we should add a comment in `dts/bindings/timer/st,stm32-lptim.yaml` on the new constraint.

Last, in the PR, addition of stm32_lp_tick_source in boards description should be done before adding the code that uses it, in order to avoid failures (on a git bisect for instance).

FRASTM commented on samples/boards/stm32/power_mgmt/blinky/boards/nucleo_wb55rg.conf:
```diff
@@ -0,0 +1,4 @@
+# set the CONFIG_SYS_CLOCK_TICKS_PER_SEC depending on the LPTIM prescaler
+# with a LPTIM1 prescaler of <1> to <8>, CONFIG_SYS_CLOCK_TICKS_PER_SEC is 4096
+# with a LPTIM1 prescaler above <16>, CONFIG_SYS_CLOCK_TICKS_PER_SEC is LSE / prescaler
+CONFIG_STM32_LPTIM_CLOCK_LSE=y
```
yes, removed from the 'samples'

FRASTM commented:
> @FRASTM Do you think we can replace the whole prescaler/frequency Kconfig based check by a runtime version ? I expect a runtime check to be much more simple. Drawback is that misconfiguration would be discovered later, but it doesn't prevent to provide a clear message to the user.

The warning message will still check CONFIG_SYS_CLOCK_TICKS_PER_SEC and the ratio CONFIG_STM32_LPTIM_CLOCK/LPTIM_PRESCALER, whatever the CONFIG_STM32_LPTIM_CLOCK is LSE or LSI ; this can be a common macro.

FRASTM commented:
changes after review

FRASTM commented:
change to ASSERT at runtime when comparing  CONFIG_SYS_CLOCK_TICKS_PER_SEC  to lptim clock frequency

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -61,12 +61,33 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
 
+static int32_t lptim_time_base;
+static uint32_t lptim_clock_freq = CONFIG_STM32_LPTIM_CLOCK;
+static uint32_t lptim_clock_presc = LPTIM_PRESCALER;
```
To simplify things a bit, can we get rid of LPTIM_PRESCALER, then ?

FRASTM replied:
LPTIM_PRESCALER  is used later to `LL_LPTIM_SetPrescaler`

FRASTM replied:
now lptim_clock_presc = DT_PROP(DT_DRV_INST(0), st_prescaler)

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -394,14 +404,29 @@ static int sys_clock_driver_init(void)
 		 * Time base = (4s * freq) - 1
 		 */
 		lptim_clock_freq = lptim_clock_freq / 2;
+		lptim_clock_presc = lptim_clock_presc * 2;
```
That would deserve a comment

FRASTM replied:
this section is removed after PR https://github.com/zephyrproject-rtos/zephyr/pull/65683

erwango reviewed (changes requested):
Looks much better.
Some few clean up changes and comments to be addressed and we should be good

FRASTM committed:
doc: releases: add the reference to stm32_lp_tick_source

Update migration guide to introduce the stm32_lp_tick_source
for stm32 device when choosing the LPTIM fo lowPower modes.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: timer: stm32 lptimer adjust the TICKS_PER_SEC to LPTIM clock

With low LPTIM freq when prescaler is set to 16 or 32,
the CONFIG_SYS_CLOCK_TICKS_PER_SEC must be reduced to
LPTIM CLOCK_/prescaler to avoid spurious timer wakeup activity.
Assert error if the CONFIG_SYS_CLOCK_TICKS_PER_SEC
is not compatible with the lptim clock freq.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase on https://github.com/zephyrproject-rtos/zephyr/pull/65683 which removes the lptim clock source divider of the stm32u5

FRASTM commented:
rebase on  9521371157c23e70cfb58b6a76aa9b65c93868a9

gautierg-st commented on samples/boards/stm32/power_mgmt/blinky/boards/b_u585i_iot02a.conf:
```diff
@@ -0,0 +1,4 @@
+# Because the stm32U5 has a pre-div on the LPTIM clock source
+# the prescaler is twice the value of the stm32_lp_tick_source
+# SYS_CLOCK_TICKS_PER_SEC must be overwritten
+CONFIG_SYS_CLOCK_TICKS_PER_SEC=1024
```
Shoudn't this be removed since the static prescaler has been removed in #65683 ?

FRASTM replied:
removed

gautierg-st commented on samples/boards/stm32/power_mgmt/blinky/boards/b_u585i_iot02a.overlay:
```diff
@@ -0,0 +1,10 @@
+  */
+&stm32_lp_tick_source {
+	st,prescaler = <16>;
+};
```
Shoudn't this be removed since the static prescaler has been removed in #65683 ?

FRASTM replied:
We can have a sample application running on this target with lptim at 2048Hz and on another platform with lptim clock at 1024Hz. Sleeping period of 2000ms is unchanged
